Support non-attention path operators in Triton

JackChuang opened the pull request:
## Motivation

As the adoption of heterogeneous hardware continues to rise—and will only accelerate in the future—it is essential for SGLang to build an execution path that operates seamlessly across different hardware platforms. Enabling CUDA-free execution is a critical step toward long-term flexibility and broader hardware compatibility, ensuring that SGLang remains adaptable to future hardware developments.

This PR is to enable CUDA-free inference serving in SGLang by introducing a Triton-based implementation for non-attention operators. SGLang's attention operators are already implemented in Triton, while non-attention operators rely on PyTorch implementations. By leveraging our Triton-based work, this change allows SGLang to achieve full CUDA-free execution without depending on PyTorch's CUDA backend.

We start with LLaMA, making all non-attention ops CUDA-free, and match the performance described in this [PyTorch blog about CUDA-free inference](https://pytorch.org/blog/cuda-free-inference-for-llms/).

## Modifications

Add a new environment flag called SGL_USE_TRITON_NON_ATTN flag for Triton-based non-attention operators.
Implement all non-attention path operators in Triton, categorized into two major types:
1. CustomOP-Based Operators: Includes RMSNorm, activation functions, and RoPE.
2. Non-CustomOP Operators: Covers other operators, including linear operators.

Usage:
```
$ SGL_USE_TRITON_NON_ATTN=1 python3 -m sglang.launch_server --model-path ...
...
[2025-03-25 19:43:06 TP0] SGL_USE_TRITON_NON_ATTN is enabled
```

## Benchmarking Results
Device: H20 * 1
Model: Llama-3.1-8B-Instruct
Api_endpoint: "/v1/completions"
<br>
Server's configuration:
All default
```
$ python3 -m sglang.launch_server --model-path /data00/models/Llama-3.1-8B-Instruct --trust-remote-code --port 8010
```


Client's configuration:
--num-prompts 32 --max-concurrency 8 --random-input-len 1000 --random-output-len 1000
Others are default = torch_compile (X) cuda_graph (O)
```
$ python3 /data00/horenc/vllm_ori/benchmarks/benchmark_serving.py --backend sglang --model /data00/models/Llama-3.1-8B-Instruct --base-url http://127.0.0.1:8010 --endpoint /v1/completions --num-prompts 32 --metric_percentiles '50,90,95,99' --max-concurrency 8 --random-input-len 1000 --random-output-len 1000 --dataset-name random --ignore-eos --seed 4
```
<br>
Results:

During the experiment, we found that warm-up time is necessary, especially for Triton. This could be part of future work. Therefore, we had the client run three consecutive times and reported the results of the last run.



[Baseline - using CUDA for non-attention operators]
```
============ Serving Benchmark Result ============
Successful requests:                     32
Benchmark duration (s):                  45.67
Total input tokens:                      32000
Total generated tokens:                  32000
Request throughput (req/s):              0.70
Output token throughput (tok/s):         700.61
Total Token throughput (tok/s):          1401.23
---------------Time to First Token----------------
Mean TTFT (ms):                          61.66
Median TTFT (ms):                        60.48
P50 TTFT (ms):                           60.48
P90 TTFT (ms):                           72.19
P95 TTFT (ms):                           72.68
P99 TTFT (ms):                           73.82
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          11.37
Median TPOT (ms):                        11.36
P50 TPOT (ms):                           11.36
P90 TPOT (ms):                           11.38
P95 TPOT (ms):                           11.39
P99 TPOT (ms):                           11.39
---------------Inter-token Latency----------------
Mean ITL (ms):                           11.37
Median ITL (ms):                         11.35
P50 ITL (ms):                            11.35
P90 ITL (ms):                            11.61
P95 ITL (ms):                            11.65
P99 ITL (ms):                            11.72
==================================================
```

[This PR - using Triton for non-attention operators]
```
============ Serving Benchmark Result ============
Successful requests:                     32
Benchmark duration (s):                  52.98
Total input tokens:                      32000
Total generated tokens:                  32000
Request throughput (req/s):              0.60
Output token throughput (tok/s):         604.04
Total Token throughput (tok/s):          1208.07
---------------Time to First Token----------------
Mean TTFT (ms):                          82.70
Median TTFT (ms):                        81.14
P50 TTFT (ms):                           81.14
P90 TTFT (ms):                           100.17
P95 TTFT (ms):                           100.70
P99 TTFT (ms):                           101.20
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          13.17
Median TPOT (ms):                        13.18
P50 TPOT (ms):                           13.18
P90 TPOT (ms):                           13.18
P95 TPOT (ms):                           13.21
P99 TPOT (ms):                           13.22
---------------Inter-token Latency----------------
Mean ITL (ms):                           13.17
Median ITL (ms):                         13.16
P50 ITL (ms):                            13.16
P90 ITL (ms):                            13.46
P95 ITL (ms):                            13.55
P99 ITL (ms):                            13.72
==================================================
```

Summary:
The performance results of Triton-only are only 13.78% lower than the current SGLang baseline. This is very close to the results reported in the [PyTorch blog on CUDA-free inference](https://pytorch.org/blog/cuda-free-inference-for-llms/) and is considered acceptable. Besides, after testing, we found that this slowdown mainly comes from linear operators.


## Checklist

- [x] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [x] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [x] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

JackChuang committed:
[Triton][Misc] Add SGL_USE_TRITON_NON_ATTN env variable and triton path in CustomOp

Add an environmental variable to control whether or not to use triton operators.

Add a forward_triton path in Custom for Triton kernel implementation.

Signed-off-by: Ho-Ren (Jack) Chuang <horenchuang@bytedance.com>
Co-authored-by: Tony Zhang <tony.zhang@bytedance.com>

JackChuang committed:
[Triton][Misc] Integrate LibEntry as a Triton kernel interface in SGLang

Integrate LibEntry as a Triton kernel interface.

Signed-off-by: Ho-Ren (Jack) Chuang <horenchuang@bytedance.com>
Co-authored-by: Tony Zhang <tony.zhang@bytedance.com>

JackChuang committed:
[Triton][Kernel] Add Triton RMSNorm and FusedAddRMSNorm

Implement forward_triton in the RMSNorm (CustomOp),
using rms_norm and fused_add_rms_norm.

Signed-off-by: Ho-Ren (Jack) Chuang <horenchuang@bytedance.com>
Co-authored-by: Tony Zhang <tony.zhang@bytedance.com>

JackChuang committed:
[Triton][Kernel] Add Triton SiLUandMul

Implement forward_triton in SiluAndMul (CustomOp), using silu_and_mul.

Signed-off-by: Ho-Ren (Jack) Chuang <horenchuang@bytedance.com>
Co-authored-by: Tony Zhang <tony.zhang@bytedance.com>

JackChuang committed:
[Triton][Kernel] Add Triton RotaryEmbedding

Implement forward_triton in RotaryEmbedding (CustomOp),
using apply_rotary_pos_emb.

Signed-off-by: Ho-Ren (Jack) Chuang <horenchuang@bytedance.com>
Co-authored-by: Tony Zhang <tony.zhang@bytedance.com>
Co-authored-by: Yichen Wang <yichen.wang@bytedance.com>
Co-authored-by: Zixuan Zhang <zixuanzhang@bytedance.com>

JackChuang committed:
[Triton][Kernel] Add Triton linear operators

Add triton linear operators mm and addmm and integrate them
with UnquantizedLinearMethod (Not CustomOp).

Signed-off-by: Ho-Ren (Jack) Chuang <horenchuang@bytedance.com>
Co-authored-by: Tony Zhang <tony.zhang@bytedance.com>

Alcanderian commented on python/sglang/srt/custom_op.py:
```diff
@@ -35,7 +39,13 @@ def forward_hpu(self, *args, **kwargs):
     def forward_cpu(self, *args, **kwargs):
         return self.forward_native(*args, **kwargs)
 
+    def forward_triton(self, *args, **kwargs):
```
Found redefinition

JackChuang replied:
Thanks for catching it! Will fix it.

Alcanderian commented on python/sglang/srt/custom_op.py:
```diff
@@ -35,7 +39,13 @@ def forward_hpu(self, *args, **kwargs):
+        return self.forward_triton(*args, **kwargs)
+
     def dispatch_forward(self):
+        if os.environ.get("SGL_USE_TRITON_NON_ATTN", "").lower() in ("true", "1"):
```
We can use `get_bool_env_var` in `srt/utils.py`

JackChuang replied:
@Alcanderian  That's great! Will leverage `get_bool_env_var`. Thanks.

Alcanderian commented on python/sglang/srt/layers/linear.py:
```diff
@@ -168,8 +547,15 @@ def apply(
     ) -> torch.Tensor:
-
-        return F.linear(x, layer.weight, bias)
+        if os.environ.get("SGL_USE_TRITON_NON_ATTN", "").lower() in ("true", "1"):
```
Should not read environment variable in forward path, we shoud move it in to initialization phase, and do not forget to use `get_bool_env_var`

JackChuang replied:
@Alcanderian I will use `get_bool_env_var ` in LinearBase.__init__ to select my new class implementing in Triton as self.quant_method. Does this sound ok?

Alcanderian replied:
Agree. That sounds more reasonable.

Alcanderian commented on python/sglang/srt/utils.py:
```diff
@@ -1733,3 +1734,185 @@ def parse_connector_type(url: str) -> str:
     return m.group(1)
+
+
+def _get_device_count():
```
We already have `get_device_count` at line1233

JackChuang replied:
@Alcanderian That's great! I will leverage the existing one. Thanks.

Alcanderian commented:
Two ideas:

1. The function `maybe_set_triton_cache_manager` in `srt/utils.py` is similar to LibEntry, so LibEntry may not be needed. But I'm not sure @zhyncs .
2. These Triton kernels should be organized into a specific folder, such as `layers/triton_ops`.

JackChuang commented:
> Two ideas:
> 
> 1. The function `maybe_set_triton_cache_manager` in `srt/utils.py` is similar to LibEntry, so LibEntry may not be needed. But I'm not sure @zhyncs .
> 2. These Triton kernels should be organized into a specific folder, such as `layers/triton_ops`.

Regarding 2., that sounds good to me. I can create a folder layers/triton_ops and move the Triton kernels there.

JackChuang committed:
[Triton][Kernel] Refactor: organize triton activation kernels into layers/triton_ops

Create a folder layers/triton_ops for placing non-attention ops.

Move Triton kernel silu_and_mul_triton() to layers/triton_ops/activation.py

Signed-off-by: Ho-Ren (Jack) Chuang <horenchuang@bytedance.com>

JackChuang committed:
[Triton][Kernel] Refactor: organize triton layernorm kernels into layers/triton_ops

Move Triton kernel rms_norm_triton() and fused_add_rms_norm_triton()
to layers/triton_ops/layernorm.py

Signed-off-by: Ho-Ren (Jack) Chuang <horenchuang@bytedance.com>

JackChuang committed:
[Triton][Kernel] Refactor: organize triton rotary_embedding kernels into layers/triton_ops

Move Triton kernel rotary_embedding_triton() to layers/triton_ops/rotary_embedding.py

Signed-off-by: Ho-Ren (Jack) Chuang <horenchuang@bytedance.com>

JackChuang committed:
[Triton][Kernel] Refactor: organize triton linear kernels into layers/triton_ops

Move Triton kernel mm_triton() and addmm_triton()
to layers/triton_ops/linear.py

Signed-off-by: Ho-Ren (Jack) Chuang <horenchuang@bytedance.com>

JackChuang commented:
Hi @Alcanderian, I’ve made the changes based on your suggestions and uploaded the v2 patchset. When you have time, I’d really appreciate it if you could take a look. Thank you~

JackChuang commented:
Hi @Alcanderian and @zhyncs ,
Just wanted to follow up on the v2 patchset I submitted — I’ve made all the changes based on your suggestions, and I’m happy to adjust anything further if needed. Totally understand you might be busy, just wanted to check in and see if there’s any update. Thanks again for your time and help!

JackChuang commented:
Thanks to @Alcanderian for the feedback!
I'll take a look and make the changes as soon I can.

JackChuang commented:
I was working on other SGLang upstream tasks recently, and I'm now getting back to this by preparing a new version of the PR:
1. Removing class LibEntry
2. Replacing torch.autograd.Function with normal function call.
3. If possible, merging my Triton kernels with the existing rmsnorm()s in layers/elementwise.py.

JackChuang committed:
[Triton][Kernel] Refactor: remove LibEntry

Signed-off-by: Ho-Ren (Jack) Chuang <horenchuang@bytedance.com>

JackChuang committed:
[Triton][Kernel] Refactor: merge duplicated triton kernel

Merge duplicated triton kernel in layers/elementwise.py

Signed-off-by: Ho-Ren (Jack) Chuang <horenchuang@bytedance.com>

JackChuang committed:
[Triton][Kernel] Refactor: remove torch.autograd.Function

Replacing Class inheriting torch.autograd.Function with normal function call.

Signed-off-by: Ho-Ren (Jack) Chuang <horenchuang@bytedance.com>

JackChuang committed:
[Triton][Kernel] Fix: add forward_triton check in RMSNorm

Since the forward() method has been overridden in RMSNorm, the
SGL_USE_TRITON_NON_ATTN check also needs to be implemented here.

Signed-off-by: Ho-Ren (Jack) Chuang <horenchuang@bytedance.com>

JackChuang commented:
Hi @Alcanderian,
I’ve addressed the points you mentioned -- points 1, 2, and 3.
Would appreciate it if you could review them when convenient. Thank you~

JackChuang commented:
Hi @Alcanderian and @zhyncs,
I've rebased to the latest main. Please check again. Thanks~

JackChuang commented:
Hi @Alcanderian and @zhyncs,
Even though I've rebased to the latest main, some files are updated very frequently, so there's still a conflict at the moment. When you have a chance, could you please help review the PR — if it looks good to merge, I'm happy to rebase again right away to resolve the conflict. Thank you so much!
